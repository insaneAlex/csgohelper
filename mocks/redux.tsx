import React from 'react';

export const createStubComponent = () => (props: any) => {
  return Object.keys(props)
    .sort()
    .map((el) => <React.Fragment key={el}>{el}</React.Fragment>);
};
