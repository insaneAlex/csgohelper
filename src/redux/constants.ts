export const INVENTORY_KEY = 'inventory';
export const FEEDBACK_KEY = 'feedback';
