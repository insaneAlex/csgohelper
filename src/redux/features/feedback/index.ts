export {
  postFeedbackSuccess,
  postFeedbackStart,
  postFeedbackError,
  FeedbackStatuses,
  feedbackReducer,
  onResetState
} from './slice';

export {feedbackStatusSelector, isFeedbackLoadingSelector} from './selectors';
