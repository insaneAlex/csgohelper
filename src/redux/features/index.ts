export {
  getItemsStart,
  getItemsError,
  itemsSelector,
  getItemsSuccess,
  inventoryReducer,
  InventoryStatuses,
  itemsFiltersSelector,
  inventoryStatusSelector,
  itemsUpdateTimeSelector
} from './inventory';

export {feedbackReducer, FeedbackStatuses, feedbackStatusSelector, postFeedbackStart, onResetState} from './feedback';
