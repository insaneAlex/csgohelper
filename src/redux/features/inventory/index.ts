export {
  getItemsStart,
  getItemsError,
  getItemsSuccess,
  inventoryReducer,
  getInitialItemsStart,
  getInitialItemsError,
  getInitialItemsSuccess
} from './slice';

export {
  itemsSelector,
  itemsErrorSelector,
  itemsFiltersSelector,
  itemsLoadingSelector,
  itemsUpdateTimeSelector
} from './selectors';
