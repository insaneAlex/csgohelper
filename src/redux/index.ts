export {
  getItemsStart,
  getItemsError,
  itemsSelector,
  getItemsSuccess,
  inventoryReducer,
  itemsFiltersSelector,
  itemsUpdateTimeSelector,
  inventoryStatusSelector
} from './features';

export type {RootState} from './store';
export type {InventoryState, SteamIDType, InventoryErrorType} from './types';
