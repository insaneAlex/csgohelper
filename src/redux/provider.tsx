import {FC, ReactNode} from 'react';
import {store} from './store';
import {Provider} from 'react-redux';

export const ReduxProvider: FC<{children: ReactNode}> = ({children}) => {
  return <Provider store={store}>{children}</Provider>;
};
