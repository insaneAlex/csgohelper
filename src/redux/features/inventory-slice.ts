import {InventoryItemType} from '@/types';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {INVENTORY_KEY} from '../constants';

type InventoryState = {
  items: InventoryItemType[];
  isLoading: boolean;
  hasError: boolean;
};
const initialState: InventoryState = {
  items: [],
  isLoading: false,
  hasError: false
};

export const inventory = createSlice({
  name: INVENTORY_KEY,
  initialState,
  reducers: {
    getItemsStart: (state: InventoryState) => {
      state.isLoading = true;
    },
    getItemsSuccess: (state: InventoryState, action: PayloadAction<InventoryItemType[]>) => {
      state.isLoading = false;
      state.items = action.payload;
    },
    getItemsError: (state: InventoryState) => {
      state.isLoading = false;
      state.hasError = true;
    }
  }
});

export const {getItemsStart, getItemsSuccess, getItemsError} = inventory.actions;

export const inventoryReducer = inventory.reducer;
