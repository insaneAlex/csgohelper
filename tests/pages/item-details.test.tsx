import {InventoryStatuses} from '@/src/redux/features';
import ItemDetailsPage from '../../pages/items/[itemId]';
import {render, screen} from '@testing-library/react';

const itemId = '123';
const query = {itemId};
const itemMock = {assetid: itemId, name: 'name1', exterior: 'super fine'};

const itemsSelectorMock = jest.fn();
const statusSelectorMock = jest.fn();
const routerReplaceMock = jest.fn();
const getFromStorageMock = jest.fn();

jest.mock('../../src/services', () => ({storage: {localStorage: {get: () => getFromStorageMock()}}}));
jest.mock('next/router', () => ({useRouter: () => ({query, replace: (a: string) => routerReplaceMock(a)})}));
jest.mock('react-redux', () => ({useDispatch: () => jest.fn(), useSelector: (selector: () => void) => selector()}));
jest.mock('../../src/redux', () => ({
  itemsSelector: () => itemsSelectorMock(),
  inventoryStatusSelector: () => statusSelectorMock(),
  getItemsStart: jest.fn()
}));

describe('ItemDetailsPage', () => {
  it('should render correctly', () => {
    itemsSelectorMock.mockImplementationOnce(() => [itemMock]);
    const {container} = render(<ItemDetailsPage />);
    expect(container).toMatchSnapshot();
  });
  describe('when loading status applied', () => {
    it('should render loader', () => {
      getFromStorageMock.mockImplementationOnce(() => '123');
      statusSelectorMock.mockImplementationOnce(() => InventoryStatuses.INIT_LOAD);
      render(<ItemDetailsPage />);
      expect(screen.getByTestId('loader')).toBeInTheDocument();
    });
  });
  describe('when has items but without selected assetid', () => {
    it('should render error message with "return to home" button', () => {
      getFromStorageMock.mockImplementationOnce(() => '123');
      itemsSelectorMock.mockImplementationOnce(() => [{...itemMock, assetid: '321'}]);
      render(<ItemDetailsPage />);
      expect(screen.getByText('There is no such item in inventory.')).toBeInTheDocument();
      expect(screen.getByText('Return to Home')).toBeInTheDocument();
    });
  });
  describe('when there is not item and no steamid', () => {
    it('should redirect to home', () => {
      getFromStorageMock.mockImplementationOnce(() => null);
      itemsSelectorMock.mockImplementationOnce(() => [{...itemMock, assetid: '321'}]);
      render(<ItemDetailsPage />);
      expect(routerReplaceMock).toHaveBeenCalledWith('/');
    });
  });
});
