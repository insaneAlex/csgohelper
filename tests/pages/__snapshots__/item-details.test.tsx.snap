// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetailsPage should render correctly 1`] = `
<div>
  <a
    href="/"
  >
    <svg
      data-testid="BackIcon"
      fill="#000000"
      height="30"
      viewBox="0 0 1024 1024"
      width="30"
    >
      <path
        d="M669.6 849.6c8.8 8 22.4 7.2 30.4-1.6s7.2-22.4-1.6-30.4l-309.6-280c-8-7.2-8-17.6 0-24.8l309.6-270.4c8.8-8 9.6-21.6 2.4-30.4-8-8.8-21.6-9.6-30.4-2.4L360.8 480.8c-27.2 24-28 64-0.8 88.8l309.6 280z"
        fill="gainsboro"
      />
    </svg>
  </a>
  <section
    class="details"
  >
    <h1>
      name1
    </h1>
    <h2
      class="exterior"
    >
      (super fine)
    </h2>
    <img
      alt="name1"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="198"
      src="https://steamcommunity-a.akamaihd.net/economy/image/undefined"
      style="color: transparent;"
      width="256"
    />
    <hr
      class="separator"
    />
  </section>
</div>
`;
